makeflow/src/dag_node.c

```diff
@@ -49,22 +50,213 @@ struct dag_node *dag_node_create(struct dag *d, int linenum)
 	return n;
 }
 
-const char *dag_node_state_name(dag_node_state_t state)
+struct dag_node_footprint *dag_node_footprint_create()
 {
-	switch (state) {
-	case DAG_NODE_STATE_WAITING:
-		return "waiting";
-	case DAG_NODE_STATE_RUNNING:
-		return "running";
-	case DAG_NODE_STATE_COMPLETE:
-		return "complete";
-	case DAG_NODE_STATE_FAILED:
-		return "failed";
-	case DAG_NODE_STATE_ABORTED:
-		return "aborted";
-	default:
-		return "unknown";
+	struct dag_node_footprint *f;
+
+	f = malloc(sizeof(struct dag_node_footprint));
+	memset(f, 0, sizeof(struct dag_node_footprint));
+
+	f->direct_children = set_create(0);
+	f->accounted = set_create(0);
+
+	f->source_size = 0;
+	f->target_size = 0;
+
+	f->terminal_files = set_create(0);
+	f->coexist_files = set_create(0);
+
+	f->residual_nodes = list_create();
+	f->residual_files = set_create(0);
+	f->residual_size = 0;
+
+	f->run_files = set_create(0);
+	f->run_footprint = 0;
+
+	f->delete_files = set_create(0);
+	f->delete_footprint = 0;
+	f->delete_run_order = list_create();
+
+	f->prog_min_files = set_create(0);
+	f->prog_min_footprint = 0;
+
+	f->prog_max_files = set_create(0);
+	f->prog_max_footprint = 0;
+	f->prog_run_order = list_create();
+
+	f->footprint_min_files = set_create(0);
+	f->footprint_min_size = 0;
+
+	f->footprint_max_files = set_create(0);
+	f->footprint_max_size = 0;
+
+	f->res = 0;
+	f->res_files = set_create(0);
+	f->wgt = 0;
+	f->wgt_files = set_create(0);
+	f->max_wgt = 0;
+	f->max_wgt_files = set_create(0);
+	f->diff = 0;
+
+	f->children_updated = 0;
+	f->size_updated = 0;
+	f->footprint_updated = 0;
+	f->terminal_updated = 0;
+
+	return f;
+}
+
+void dag_node_footprint_delete(struct dag_node_footprint *f);
+
+void dag_node_delete(struct dag_node *n)
+{
+	hash_table_delete(n->variables);
+
+	itable_delete(n->remote_names);
+	hash_table_delete(n->remote_names_inv);
+
+	set_delete(n->descendants);
+	set_delete(n->ancestors);
+
+	list_delete(n->source_files);
+	list_delete(n->target_files);
+
+	if(n->footprint)
+		dag_node_footprint_delete(n->footprint);
+
+	rmsummary_delete(n->resources_requested);
+	if(n->resources_measured)
+		rmsummary_delete(n->resources_measured);
+
+	free(n);
+}
+
+void dag_node_footprint_delete(struct dag_node_footprint *f)
+{
+	set_delete(f->direct_children);
+	set_delete(f->accounted);
+
+	set_delete(f->terminal_files);
+	set_delete(f->coexist_files);
+
+	list_delete(f->residual_nodes);
+	set_delete(f->residual_files);
+
+	set_delete(f->run_files);
+
+	set_delete(f->delete_files);
+	list_delete(f->delete_run_order);
+
+	set_delete(f->prog_min_files);
+
+	set_delete(f->prog_max_files);
+	list_delete(f->prog_run_order);
+
+	set_delete(f->footprint_min_files);
+	set_delete(f->footprint_max_files);
+
+	set_delete(f->res_files);
+	set_delete(f->wgt_files);
+	set_delete(f->max_wgt_files);
+
+	free(f);
+}
+
+int dag_node_compare(struct dag_node *node1, struct dag_node *node2, dag_node_sort_t type)
```

Comment:
In fact, if you are only using this function for the comparators below, I would eliminate this function and add the code to the corresponding comparators.